Separate out primitives (#4640)
Qiskit/documentation: docs/guides/executor-input-output.mdx

Reviewer: Update metadata title to the page title

---
title: Inputs and outputs
description: Understand the inputs and outputs to the Executor primitive.

---

# Executor inputs and outputs

The inputs and output of the Executor primitive are very different from those of the Sampler and Estimator primitives. As part of the [directed execution model,](/docs/guides/directed-execution-model) the Executor primitive helps provide more flexibility when customizing your error mitigation workflow.  This primitive takes a `QuantumProgram` as input, and outputs a Qiskit Runtime job, which is then run on an IBM&reg; quantum computer.

<span id="programs"></span>
## Inputs: Quantum programs

The input to an Executor primitive is a  [`QuantumProgram`](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgram.html#qiskit_ibm_runtime.quantum_program.QuantumProgram), which is an iterable of a
[`qiskit_ibm_runtime.quantum_program.QuantumProgramItem`.](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgramItem.html) Each of these items represents a

> different task for the Executor to perform. Typically, each item owns one of the following options:

Reviewer: ```suggestion
different task for Executor to perform. Typically, each item owns one of the following options:
```

* A `qiskit.circuit.QuantumCircuit` with static, non-parametrized gates
* A parametrized `qiskit.circuit.QuantumCircuit` with an array of parameter values
* A parametrized `qiskit.circuit.QuantumCircuit` with a `samplomatic.samplex.Samplex` to generate randomize arrays of parameter values

> Each of these items, including instructions to add them to a `QuantumProgram` is explained in more detail below.

Reviewer: ```suggestion
Each of these items, including instructions to add them to a `QuantumProgram`, is explained in more detail below.
```

# Initialize an empty program
program = QuantumProgram(shots=1024)

# Initialize and transpile a circuit
circuit = ...
...
isa_circuit = preset_pass_manager.run(circuit)

# Append the circuit to the program
program.append_circuit_item(isa_circuit)
```

Next, append a second item that contains a parametrized `qiskit.circuit.QuantumCircuit` and an array containing 10 sets of parameter values. This amounts to a circuit task requiring a total of 10240 shots (1024 per set of parameter values).

```python
# Append the circuit and the parameter value to the program
program.append_circuit_item(
    isa_circuit,
    circuit_arguments=np.random.rand(10, 3),  # 10 sets of parameter values
)
```

Finally, append a parametrized `qiskit.circuit.QuantumCircuit` and a `samplomatic.samplex.Samplex`, which is responsible for generating randomized sets of parameters for the given circuit. As part of the `samplomatic.samplex.Samplex` arguments, `10` sets of parameters for the parametric gates in the original circuit are provided. The `shape` request argument requests an extension of the implicit shape defined by the `samplomatic.samplex.Samplex` arguments. In particular, by setting `shape`

> to `(2, 14, 10)` we request to randomize each of the `10` sets of parameters `28` times, and to arrange the randomized parameter sets in an array of be arranged in an array of shape `(2, 14, 10)`.

Reviewer: ```suggestion
to `(2, 14, 10)` we request to randomize each of the `10` sets of parameters `28` times, and to arrange the randomized parameter sets in an array of shape `(2, 14, 10)`.
```

        "parameter_values": np.random.rand(10, 3),
    },
    shape=(2, 14, 10),
)
```

## Outputs

Executor's output is a [`qiskit_ibm_runtime.quantum_program.QuantumProgramResult`,](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgramResult.html#qiskit_ibm_runtime.quantum_program.QuantumProgramResult) which is an iterable. It contains one
item per circuit task, and the items are in the same order as those in the program. Each of
these items is a dictionary where the keys are strings and the values are of type `np.ndarray`, with elements of type `bool`. 

In the previous example, the result contains three items:

### Result item 1. 
The first item in `result` is the result of running the first task in the program, the circuit with static gates. 

It contains a single key, `'meas'`, corresponding to the name of the input circuit's 
classical register, mapped to the results collected for this
classical register.  The information is stored in an `np.ndarray` of shape `(shots, register bits)`, for the above example, `(1024, 3)`. 

The following code illustrates how to access this information:

```python    

> # Access the results of the classical register of task #0

Reviewer 2: It's confusing that it's titled `Result item 1` when the task is #0

---
title: Executor inputs and outputs
description: Understand the inputs and outputs to the Executor primitive.

---

# Executor inputs and outputs

The Executor primitive is part of the [directed execution model](/docs/guides/directed-execution-model), which helps provide more flexibility when customizing an error mitigation workflow. This workflow introduces some new terminology:

- A **program** is a list of program items, where each item is either a circuit item or a samplex item.
- A **samplex item** is a pair that consists of a template circuit and a samplex.

> - **Samplex** is the core type of the [Samplomatic library.](/docs/guides/directed-execution-model#samplomatic)

Reviewer 3: I don't think introducing these new terminologies here helps much given there is no context nor examples. I think going into more details in the Inputs section is sufficient.

> The inputs and output of the Executor primitive are very different from those of the [Sampler](/docs/guides/sampler-input-output) and [Estimator](/docs/guides/estimator-input-output) primitives. This primitive takes a `QuantumProgram` as input, and then outputs a Qiskit Runtime job, which is then run on an IBM&reg; quantum computer.

Reviewer 3: Here we should explain a bit why we dropped Sampler/Estimator input/output format and used something different - .e.g 

> The inputs and output of the Executor primitive are very different from those of the [Sampler](https://qiskit.github.io/documentation/pr-4640/docs/guides/sampler-input-output) and [Estimator](https://qiskit.github.io/documentation/pr-4640/docs/guides/estimator-input-output) primitives. 
> 
> Instead of taking a list of PUBs as the input, the Executor takes a `QuantumProgram` which contains a list of `QuantumProgramItem`. These container classes give you more flexibility than a PUB, which is a simple tuple data structure.
> 
> The output of the Executor is a `QuantumProgramResult`, which is an iterable and contains one element for each input `QuantumProgramItem`. Each element is a dictionary whose keys correspond to the names of the classical registers in the input circuits (among others), so you no longer need to memorize these names like you did with Sampler output.

<span id="programs"></span>
## Inputs: Quantum programs

The input to an Executor primitive is a  [`QuantumProgram`](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgram.html#qiskit_ibm_runtime.quantum_program.QuantumProgram), which is an iterable of a
[`qiskit_ibm_runtime.quantum_program.QuantumProgramItem`](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgramItem.html). Each of these items represents a
different task for Executor to perform. Typically, each item owns one of the following options:

* A `qiskit.circuit.QuantumCircuit` with static, non-parametrized gates
* A parametrized `qiskit.circuit.QuantumCircuit` with an array of parameter values

> * A parametrized `qiskit.circuit.QuantumCircuit` with a `samplomatic.samplex.Samplex` to generate randomized parameter value arrays

Reviewer 3: ```suggestion
* A parametrized `qiskit.circuit.QuantumCircuit` with a `samplomatic.samplex.Samplex`. A Samplex is a core type of the [Samplomatic library](https://github.com/Qiskit/samplomatic) and encodes all the information about the randomization process itself.
```

"to generate randomized parameter value arrays" sounds to me it's also generating the parameters for the original parametrized circuit.

Each of these items, including instructions to add them to a `QuantumProgram`, is explained in more detail below.

> ### Example: Initialize a `QuantumProgram` and append circuits

Reviewer 3: ```suggestion
### Example: Create a `QuantumProgram` with three different tasks
```

"append circuits" to me is just about appending a circuit item (no Samplex). So I was expecting to see a different example for Samplex lol.

I'd also break this example into 3 sections, similar to how it's done for the output example.

> You must first initialize your quantum program, then append circuits to the program, as shown in the following examples. 

Reviewer 3: ```suggestion
You must first initialize your quantum program, then append program items to it using either `append_circuit_item` or `append_samplex_item` (if samplex is present), as shown in the following examples.
```

> The below cell initializes a `QuantumProgram` and specifies that it should run 1024 shots for every configuration of each item in the program. Then it appends a version of the target circuit with parameters, transpiled according to the backend's instruction set architecture (ISA).

Reviewer 3: We should note that unlike Sampler, a `QuantumProgram` only takes a single `shot` value. You'd need to use a separate `QuantumProgram` (separate job) if you want a different shot value.

```python
from qiskit.transpiler import generate_preset_pass_manager
from qiskit_ibm_runtime.quantum_program import QuantumProgram
import numpy as np
from samplomatic import build
from samplomatic.transpiler import generate_boxing_pass_manager

# Initialize an empty program
program = QuantumProgram(shots=1024)

# Initialize and transpile a circuit
circuit = ...
...
isa_circuit = preset_pass_manager.run(circuit)

# Append the circuit to the program
program.append_circuit_item(isa_circuit)
```

Next, append a second item that contains a parametrized `qiskit.circuit.QuantumCircuit` and an array containing 10 sets of parameter values. This amounts to a circuit task requiring a total of 10240 shots (1024 per set of parameter values).

```python
# Append the circuit and the parameter values to the program
program.append_circuit_item(
    parametrized_isa_circuit,

>     circuit_arguments=np.random.rand(10, 3),  # 10 sets of parameter values

Reviewer 3: If we are not going to show the full example of how the circuit was constructed, we should at least mention that it contains 3 parameters (i.e. where this `3` came from)

)
```

> Finally, append a parametrized `qiskit.circuit.QuantumCircuit` and a `samplomatic.samplex.Samplex`, which is responsible for generating randomized sets of parameters for the given circuit. As part of the `samplomatic.samplex.Samplex` arguments, 10 sets of parameters for the parametric gates in the original circuit are provided. The `shape` request argument requests an extension of the implicit shape defined by the `samplomatic.samplex.Samplex` arguments. 

Reviewer 3: This example would be really confusing to a new user of Executor. For example, what does `generate_boxing_pass_manager` do? Why do I need Samplex when I didn't need it in the last 2 examples? What does the last sentence mean?

Reviewer 4: I would go with:
```suggestion
Circuit items are executed without any sort of randomization. On the contrary, samplex items allows users to specify how to randomize their content. In the next cell, we show how to generate a samplex items for the GHZ circuit above that adds twirling to all the gates and measurements in it.
```

Then in the snippet, you can add inline comments for the various steps. See my comments below

In particular, setting `shape`

> to `(2, 14, 10)` specifies to randomize each of the 10 sets of parameters 28 (`2 x 14`) times, and to arrange the randomized parameter sets in an array of shape `(2, 14, 10)`.

Reviewer 3: If I want 28 randomizations, can I just do `shape=(28, 10)`? If not, why not?

Reviewer 4: yes

> See the samplomatic [API](https://github.com/Qiskit/samplomatic/) documentation for full details about `samplomatic.samplex.Samplex` and its arguments.

Reviewer: ```suggestion
See the Samplomatic [API](https://github.com/Qiskit/samplomatic/) documentation for full details about `samplomatic.samplex.Samplex` and its arguments.
```

template, samplex = build(boxed_circuit)

# Append the template and samplex as a samplex item
program.append_samplex_item(
    template,
    samplex=samplex,
    samplex_arguments={
        # the arguments required by the samplex.sample method
        "parameter_values": np.random.rand(10, 3),
    },
    shape=(2, 14, 10),
)
```

## Outputs

Executor's output is a [`qiskit_ibm_runtime.quantum_program.QuantumProgramResult`,](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgramResult.html#qiskit_ibm_runtime.quantum_program.QuantumProgramResult) which is an iterable. It contains one
item per circuit task, and the items are in the same order as those in the program. Each of
these items is a dictionary where the keys are strings and the values are of type `np.ndarray`, with elements of type `bool`. 

The result for the previous example contains three items:

### Result item 0
The first item in `result` is the result of running the first task in the program, the circuit with static gates. 

It contains a single key, `'meas'`, corresponding to the name of the input circuit's 
classical register, mapped to the results collected for this

> classical register.  The information is stored in an `np.ndarray` of shape `(shots, register bits)`, which is `(1024, 3)` for the above example.

Reviewer 3: I know it's implied, but the "above example" doesn't actually show or mention the circuit has a classical register with 3 bits.

---
title: Executor inputs and outputs
description: Understand the inputs and outputs to the Executor primitive.

---

# Executor inputs and outputs

The Executor primitive is part of the [directed execution model](/docs/guides/directed-execution-model), which helps provide more flexibility when customizing an error mitigation workflow. This workflow introduces some new terminology:

The inputs and output of the Executor primitive are very different from those of the [Sampler](/docs/guides/sampler-input-output) and [Estimator](/docs/guides/estimator-input-output) primitives. 

Instead of taking a list of PUBs as the input, Executor takes a `QuantumProgram`, which contains a list of `QuantumProgramItem` objects. These container classes give you more flexibility than a PUB, which is a simple tuple data structure.

> Executor's output is a `QuantumProgramResult`, which is an iterable and contains one element for each input `QuantumProgramItem`. Each element is a dictionary whose keys correspond the classical registers' names in the input circuits (among others), so you no longer need to memorize these names like you did with Sampler output.

Reviewer 4: Consider giving a few hints about what these new objects are, to make them look less scary
```suggestion
The Executor primitive is part of the [directed execution model](/docs/guides/directed-execution-model), which helps provide more flexibility when customizing an error mitigation workflow. 

The inputs and output of the Executor primitive are very different from those of the [Sampler](/docs/guides/sampler-input-output) and [Estimator](/docs/guides/estimator-input-output) primitives. For example, instead of taking a list of PUBs as the input, Executor takes a `QuantumProgram`, which contains a list of `QuantumProgramItem` objects. `QuantumProgramItem`s can be of two types:
- `CircuitItem`, which typically stores a circuit and its parameters.
- `SamplexItem`, which typically stores a circuit and a `Samplex` object, capable of generating (randomized) sets of parameters at runtime.
These container classes give you more flexibility than a PUB, which is a simple tuple data structure. Indeed, they allow you to specify where and how a set of circuit instructions should be randomized before execution, for example to perform twirling or injecting noise.

Executor's output is a `QuantumProgramResult`, which is an iterable and contains one element for each input `QuantumProgramItem`. Each element is a dictionary whose keys correspond the classical registers' names in the input circuits (among others), so you no longer need to memorize these names like you did with Sampler output.
```

This primitive takes a `QuantumProgram` as input, and then outputs a Qiskit Runtime job, which is then run on an IBM&reg; quantum computer.

<span id="programs"></span>
## Inputs: Quantum programs

The input to an Executor primitive is a  [`QuantumProgram`](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgram.html#qiskit_ibm_runtime.quantum_program.QuantumProgram), which is an iterable of a

> [`qiskit_ibm_runtime.quantum_program.QuantumProgramItem`](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgramItem.html). Each of these items represents a

Reviewer 4: Reader's feedback: I prefer how this renders:
```suggestion
[`QuantumProgramItem`](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgramItem.html). Each of these items represents a
```

### Example: Create a `QuantumProgram` with three different tasks

You must first initialize your quantum program, then append program items to it using either `append_circuit_item` or `append_samplex_item` (if samplex is present), as shown in the following examples.

The below cell initializes a `QuantumProgram` and specifies that it should run 1024 shots for every configuration of each item in the program. Then it appends a version of the target circuit with parameters, transpiled according to the backend's instruction set architecture (ISA).

<Admonition type="note">
Unlike with Sampler, a `QuantumProgram` takes only a single shot value. If you want a different shot value, you need a separate `QuantumProgram`, which would be a separate job.
</Admonition>

```python
from qiskit.transpiler import generate_preset_pass_manager
from qiskit_ibm_runtime.quantum_program import QuantumProgram
import numpy as np
from samplomatic import build
from samplomatic.transpiler import generate_boxing_pass_manager

# Initialize an empty program
program = QuantumProgram(shots=1024)

# Initialize and transpile a 3-qubit quantum circuit with 3 parameters.

> circuit = ...

Reviewer 4: A few comments:

1. I think that three examples are excessive, I'd just go with the parametric circuit directly (cut this one without params, it buys us nothing).
2. As a user, I like to copy paste cells that work. So, consider adding a simple GHZ circuit:
```suggestion
circuit = QuantumCircuit(3)
circuit.h(0)
circuit.cx(0, 1)
circuit.cx(1, 2)
circuit.rz(Parameter("theta"), 0)
circuit.rz(Parameter("phi"), 1)
circuit.rz(Parameter("lam"), 2)
```
3. You'd then have to generate the params and add them in `program.append_circuit_item(isa_circuit)`

...

> # When the circuit is measured in this way, a 3-bit classical register is added.

Reviewer 4: ```suggestion
# `measure_all` adds a 3-bit classical register named "meas"
```

circuit.measure_all()

isa_circuit = preset_pass_manager.run(circuit)

# Append the circuit to the program
program.append_circuit_item(isa_circuit)
```

Next, append a second item that contains a parametrized `qiskit.circuit.QuantumCircuit` and an array containing 10 sets of parameter values. This amounts to a circuit task requiring a total of 10240 shots (1024 per set of parameter values).

```python
# Append the circuit and the parameter values to the program
program.append_circuit_item(
    parametrized_isa_circuit,
    circuit_arguments=np.random.rand(10, 3),  # 10 sets of parameter values
)
```

Finally, append a parametrized `qiskit.circuit.QuantumCircuit` and a `samplomatic.samplex.Samplex`, which is responsible for generating randomized sets of parameters for the given circuit. As part of the `samplomatic.samplex.Samplex` arguments, 10 sets of parameters for the parametric gates in the original circuit are provided. The `shape` request argument requests an extension of the implicit shape defined by the `samplomatic.samplex.Samplex` arguments. 

> In particular, setting `shape`

Reviewer 4: maybe remove this? seems out of place, and can just comment in the snippet below

to `(2, 14, 10)` specifies to randomize each of the 10 sets of parameters 28 (`2 x 14`) times, and to arrange the randomized parameter sets in an array of shape `(2, 14, 10)`.

See the Samplomatic [API](https://github.com/Qiskit/samplomatic/) documentation for full details about `samplomatic.samplex.Samplex` and its arguments.

```python
# Transpile the circuit and group gates and measurements into annotated boxes
preset_pass_manager = generate_preset_pass_manager(
    backend=backend, optimization_level=0
)

> preset_pass_manager.post_scheduling = generate_boxing_pass_manager(

Reviewer 4: example of inline comment (please reword as you see fit)
```suggestion
# the boxing pass manager is a convenient way to group gates
# and measurements into boxes and request that they be annotated; in this example, we request that these boxes
# be annotated with a`Twirl` annotation
preset_pass_manager.post_scheduling = generate_boxing_pass_manager(
```

    enable_gates=True,
    enable_measures=True,
)
boxed_circuit = preset_pass_manager.run(circuit)

> # Build the template and the samplex

Reviewer 4: ```suggestion
# Build the template and the samplex. The template circuit is a
# parametric circuit, and the samplex is an object that can generate
# randomized sets of parameters at runtime to perform twirling
```

template, samplex = build(boxed_circuit)

# Append the template and samplex as a samplex item
program.append_samplex_item(
    template,
    samplex=samplex,
    samplex_arguments={
        # the arguments required by the samplex.sample method
        "parameter_values": np.random.rand(10, 3),
    },

>     shape=(2, 14, 10),

Reviewer 4: ```suggestion
    shape=(28, 10),   # 28 randomizations 
```

---
title: Executor inputs and outputs
description: Understand the inputs and outputs to the Executor primitive.

---

# Executor inputs and outputs

The Executor primitive is part of the [directed execution model](/docs/guides/directed-execution-model), which helps provide more flexibility when customizing an error mitigation workflow. 

The inputs and output of the Executor primitive are very different from those of the [Sampler](/docs/guides/sampler-input-output) and [Estimator](/docs/guides/estimator-input-output) primitives. For example, instead of taking a list of PUBs as the input, Executor takes a `QuantumProgram`, which contains a list of `QuantumProgramItem` objects. `QuantumProgramItem`s can be of two types:
- `CircuitItem`, which typically stores a circuit and its parameters.
- `SamplexItem`, which typically stores a circuit and a `Samplex` object, capable of generating (randomized) sets of parameters at runtime.

> These container classes give you more flexibility than a PUB, which is a simple tuple data structure. Indeed, they allow you to specify where and how a set of circuit instructions should be randomized before execution, for example to perform twirling or injecting noise.

Reviewer 4: This sounds a bit better than my previous suggestion
```suggestion
- `SamplexItem`, which typically stores a circuit, its parameters, and a `Samplex` object. The samplex is capable of generating randomized sets of parameters at runtime, for example to perform twirling or inject noise.

These container classes give you more flexibility than a PUB, which is a simple tuple data structure. For example, they allow you to specify where and how a set of circuit instructions should be randomized before execution.
```

# be annotated with a`Twirl` annotation
preset_pass_manager.post_scheduling = generate_boxing_pass_manager(
    enable_gates=True,
    enable_measures=True,
)
boxed_circuit = preset_pass_manager.run(circuit)

# Build the template and the samplex. The template circuit is a
# parametric circuit, and the samplex is an object that can generate
# randomized sets of parameters at runtime to perform twirling
template, samplex = build(boxed_circuit)

# Append the template and samplex as a samplex item
program.append_samplex_item(
    template,
    samplex=samplex,
    samplex_arguments={
        # the arguments required by the samplex.sample method
        "parameter_values": np.random.rand(10, 3),
    },
    shape=(28, 10),   # 28 randomizations 
)
```

## Outputs

Executor's output is a [`qiskit_ibm_runtime.quantum_program.QuantumProgramResult`,](https://qiskit.github.io/qiskit-ibm-runtime/stubs/qiskit_ibm_runtime.quantum_program.QuantumProgramResult.html#qiskit_ibm_runtime.quantum_program.QuantumProgramResult) which is an iterable. It contains one
item per circuit task, and the items are in the same order as those in the program. Each of

> these items is a dictionary where the keys are strings and the values are of type `np.ndarray`, with elements of type `bool`. 

Reviewer 4: Not always true. For example, if you set `program.meas_type = "kerneled"`, they will be complex numbers. I would leave it like this:
```suggestion
these items is a dictionary where the keys are strings and the values are of type `np.ndarray`. 
```
